File: durmitor/src/main/java/ql/helpers/symboltable/Symbol.java
package ql.helpers.symboltable;

import ql.ast.expression.Identifier;
import ql.ast.type.Type;

public class Symbol {

    private Identifier id;
    private Type type;
    
    public Symbol(Identifier id, Type type) {
        
        this.id         = id;
        this.type       = type;
    }
    
    public Identifier getIdentifier() {
        return id;
    }
    
    public Type getType() {
        return type;
    }
    
    public boolean equals(Symbol symbol) {
        
        return  id.getName().equals(symbol.getIdentifier().getName()) &&
                type.equals(symbol.getType());
    }
    
}

File: DogeQL/src/main/kotlin/doge/data/symbol/Symbol.kt
package doge.data.symbol

import doge.ast.node.expression.Expression
import doge.data.question.SymbolType
import doge.data.value.BaseSymbolValue

class Symbol(type: SymbolType, val expression: Expression?) {

    var value: BaseSymbolValue = type.getDefaultInstance()

    fun evaluate(symbolTable: SymbolTable) {
        expression?.let {
            value = expression.evaluate(symbolTable)
        }
    }

    fun assign(value: BaseSymbolValue) {
        this.value = value
    }

}